### PR TITLE
Feature/storage power actor cboring

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Storage power actor state with CBOR encoding/decoding.
According to specs-actor.

### CODE REVIEW

File: core/adt/impl/balance_table_hamt.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  OUTCOME_TRY(new_root, hamt_->flush());
+  root = new_root;
+  return fc::outcome::success();
+}
+
+fc::outcome::result<TokenAmount> BalanceTableHamt::subtractWithMinimum(
+    const Address &key, const TokenAmount &amount, const TokenAmount &floor) {
+  OUTCOME_TRY(balance, hamt_->getCbor<TokenAmount>(encodeToByteString(key)));
+  TokenAmount available = balance < floor ? 0 : balance - floor;
+  TokenAmount sub = available < amount ? available : amount;
```
Comment (turuslan) [inline core/adt/impl/balance_table_hamt.cpp]: `std::min(available, amount)`

Comment (Alexey-N-Chernyshov) [reply core/adt/impl/balance_table_hamt.cpp]: fixed

File: core/adt/balance_table_hamt.hpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#include "storage/hamt/hamt.hpp"
+#include "storage/ipfs/datastore.hpp"
+
+namespace fc::adt {
+
+  using primitives::BigInt;
+  using primitives::address::Address;
+  using storage::hamt::Hamt;
+  using storage::ipfs::IpfsDatastore;
+  using TokenAmount = BigInt;
```
Comment (artyom-yurin) [inline core/adt/balance_table_hamt.hpp]:
```suggestion
  using TokenAmount = primitives::BigInt;
```
And remove `
using primitives::BigInt
`

File: core/adt/balance_table_hamt.hpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+     * Removes from table
+     * @param key addrss
+     * @return error in case of failure
+     */
+    outcome::result<void> remove(const Address &key);
+
+    /**
+     * HAMT CID of root
+     */
+    CID root;
```
Comment (artyom-yurin) [inline core/adt/balance_table_hamt.hpp]: Why not private?

File: core/adt/impl/balance_table_hamt.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#include "primitives/address/address_codec.hpp"
+
+using fc::adt::BalanceTableHamt;
+using fc::adt::TokenAmount;
+using fc::primitives::address::encodeToByteString;
+
+BalanceTableHamt::BalanceTableHamt(std::shared_ptr<IpfsDatastore> datastore,
+                                   const CID &new_root)
+    : hamt_{datastore} {
+  root = new_root;
```
Comment (artyom-yurin) [inline core/adt/impl/balance_table_hamt.cpp]: Maybe initialise in initialise list

File: core/adt/balance_table_hamt.hpp (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply core/adt/balance_table_hamt.hpp]: it is used in CBOR operators >>, <<.

File: core/adt/impl/balance_table_hamt.cpp (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply core/adt/impl/balance_table_hamt.cpp]: fixed

File: core/adt/balance_table_hamt.hpp (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply core/adt/balance_table_hamt.hpp]: done

File: core/vm/actor/builtin/storage_power/storage_power_actor_state.hpp [hunk 5]
```diff
@@ -149,49 +233,106 @@ namespace fc::vm::actor::builtin::storage_power {
...
 
-    std::shared_ptr<Indices> indices_;
+  };
 
-    std::shared_ptr<crypto::randomness::RandomnessProvider>
-        randomness_provider_;
+  /**
+   * CBOR serialization of Claim
+   */
+  template <class Stream,
```
Comment (turuslan) [inline core/vm/actor/builtin/storage_power/storage_power_actor_state.hpp]: `CBOR_ENCODE(Claim, claim) {` (from "codec/cbor/streams_annotation.hpp")

Comment (Alexey-N-Chernyshov) [reply core/vm/actor/builtin/storage_power/storage_power_actor_state.hpp]: done

**Pull request merged**